### PR TITLE
[CLOUD-2788 / CLOUD-2546] - support for upstream KUBE_PING and DNS_PING

### PR DESCRIPTION
**Pull request by luck3y**
https://issues.jboss.org/browse/CLOUD-2788
https://issues.jboss.org/browse/CLOUD-2546
Signed-off-by: Ken Wills <kwills@redhat.com>

### CODE REVIEW

**Comment by luck3y** [comment]
@rcernich FYI, while looking at this, I decided to add support for upstream dns.DNS_PING as well. It works fine locally. I need to to some additional research regarding the use of .svc.cluster.local as a mandatory suffix in DNS, as its not clear to me yet whether is is a K8S requirement or not.

**Comment by luck3y** [comment]
Reviewing https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/ it suggests that .svc.cluster.local is a K8S requirement, so I my concern these is seemingly unfounded.

**Comment by luck3y** [comment]
One other note / reminder on this. The templates all currently use openshift.DNS_PING, but the ha.sh actual default was openshift.KUBE_PING. This default was changed to kubernetes.KUBE_PING as the title of the ticket was 'Switch to using upstream KUBE_PING." I could go either way on the defaults though, as presumably once we're happy with this we can deprecate the openshift.* ones and drop them at some point.

**Comment by rcernich** [comment]
The default is that way because before the env was added, the templates used KUBE_PING, so this was set that way to support backward compatibility with existing deployments.

As for hard coding cluster.local, I'll leave it up to you guys.  I think the main issue is that DnsContext does not apply search suffixes, thus the behavior between using DnsContext will differ from things like nslookup/dig, and even InetAddress.getByName(), all of which apply the search suffixes in /etc/resolve.conf.

**Comment by luck3y** [comment]
@rcernich Yeah, dns.DNS_PING assumes that you'll be using .svc.cluster.local for the domain, as does the K8S docs. Is it likely that people have ping services running with different domain names?

@slaskawi WDYT, is assuming .svc.cluster.local a problem?

**Comment by rcernich** [comment]
The other addition to resolve.conf is that it also adds the namespace, so if you have a service named foo, you can use InetAddress.getByName("foo") and it will return the service's address, without any additional magic.  Not sure why you prefer DnsContext over InetAddress.

**Comment by luck3y** [comment]
@rcernich I agree with you regarding /etc/resolv.conf, it seems to me that this is a generally more flexible approach.

File: jboss-eap-config-jgroups/added/launch/ha.sh [hunk 1]
```diff
@@ -86,6 +93,53 @@ get_socket_binding_for_ping() {
...
+    local service_name="${3}"
+    local ping_namespace="${4}"
+    local socket_binding="${5}"
+    local config
+    # for DNS_PING, the record is my-port-name._my-port-protocol.my-svc.my-namespace
+    config="<protocol type=\"${ping_protocol}\" ${socket_binding}>"
+    if [ "${ping_protocol}" = "dns.DNS_PING" ]; then
+        if [ -z "${KUBERNETES_SERVICE_PORT_DNS}" ]; then
+            log_error "KUBERNETES_SERVICE_PORT_DNS is required for dns.DNS_PING."
+        fi
```
Comment (bstansberry) [inline jboss-eap-config-jgroups/added/launch/ha.sh]: This 'if' block conflicts with the one at L130. There it logs a WARN that it's using a default. Here it logs an ERROR that the var must be set, which isn't true as we default it.

File: jboss-eap-config-jgroups/added/launch/ha.sh [hunk 2]
```diff
@@ -86,6 +93,53 @@ get_socket_binding_for_ping() {
...
+        kdns_port=${KUBERNETES_SERVICE_PORT_DNS:-53}
+        if [ -z "${KUBERNETES_SERVICE_PORT_DNS}" ]; then
+            log_warn "KUBERNETES_SERVICE_PORT_DNS is not set, using default value of 53."
```
Comment (bstansberry) [inline jboss-eap-config-jgroups/added/launch/ha.sh]: Is 53 the typical value?  If so, is this a WARN or just an INFO?

File: jboss-eap-config-jgroups/added/launch/ha.sh [hunk 3]
```diff
@@ -86,6 +93,53 @@ get_socket_binding_for_ping() {
...
+        fi
+        kdns_proto=$(eval echo "\$KUBERNETES_PORT_"$kdns_port"_UDP_PROTO")
+        if [ -z "${kdns_proto}" ]; then
+            log_error "Unable to determine KUBERNETES_SERVICE_PORT_DNS configured protocol, KUBERNETES_PORT_"$kdns_port"_UDP_PROTO is not set."
```
Comment (bstansberry) [inline jboss-eap-config-jgroups/added/launch/ha.sh]: Below we use 'udp' as the default if this is not set, so is this more an INFO?

**Comment by bstansberry** [comment]
Questions about log levels aside, this looks fine.

Do you think the questions about DirContext vs /etc/resolv.conf will affect this change? For example, change in dns.DNS_PING impl would require a change in how this configuration is done.

File: jboss-eap-config-jgroups/added/launch/ha.sh (hunk 2 above)
Comment (luck3y) [reply jboss-eap-config-jgroups/added/launch/ha.sh]: Yeah, INFO makes more sense.

**Comment by luck3y** [comment]
I've tried to construct the dns_query portion to be independent of any external upstream change, like this:

<property name="dns_query">8888._udp.myapp.myproject</property>

Currently the upstream dns.DNS_PING will append .cluster.svc.local to this value and use that for the lookup (openshift.DNS_PING handles these details internally, so there is no additional configuration).

If / when upstream drops the .cluster.svc.local suffix, then this should just use (via the JVM's linking against libc etc) whatever search domains are configured in /etc/resolv.conf.

**Comment by rcernich** [comment]
`.myproject` is the tricky bit, as that is environment specific, which complicates configuration, as the namespace needs to be added to the service name.  Do you need the `8888._udp` portion of the name?

**Comment by rcernich** [comment]
You can use this as a reference for the types of "names" supported: https://github.com/openshift/origin/blob/master/pkg/dns/serviceresolver.go#L76

**Comment by luck3y** [comment]
> `.myproject` is the tricky bit, as that is environment specific, which complicates configuration, as the namespace needs to be added to the service name. Do you need the `8888._udp` portion of the name?

This is a good point, the port / protocol is really only for SRV record lookup, and the upstream DNS_PING currently only uses A records (https://github.com/belaban/JGroups/blob/master/src/org/jgroups/protocols/dns/DNS_PING.java#L34). 

So I should change this to not have the port / proto imo, just something like:

> \<property name="dns_query"\>${ping_service_name}.${ping_service_namespace}\</property\>

I'll update this and test.

**Comment by slaskawi** [comment]
@luck3y @rcernich Comments below:

> @slaskawi WDYT, is assuming .svc.cluster.local a problem?

I already created a ticket to drop the suffix (https://issues.jboss.org/browse/JGRP-2292).

> As for hard coding cluster.local, I'll leave it up to you guys. I think the main issue is that DnsContext does not apply search suffixes, thus the behavior between using DnsContext will differ from things like nslookup/dig, and even InetAddress.getByName(), all of which apply the search suffixes in /etc/resolve.conf.

This can be easily fixed. At the time of writing this protocol, my main focus was to connect to a specified DNS Server (not necessarily the one used by your OS). That's why you can specify the DNS Server address there. 

Now, I think we opened a new "Pandora Box" with different set of use cases. In order to use a local OS resolver, I think we should add a new [DNSResolver](https://github.com/belaban/JGroups/blob/master/src/org/jgroups/protocols/dns/DNSResolver.java) implementation. @rcernich, I think you have some experience with this kind of code. May I ask you for a contribution?

**Comment by rcernich** [comment]
@luck3y, A records should work just fine to locate the pod endpoints, but you need to make sure the service is created with `clusterIP: None`.  This is a "headless" service and the pod ip's should get served up for the A records (as opposed to the service ip, when the service is not headless).

@slaskawi, not sure what you're asking.  The openshift.DNS_PING implementation simply uses InetAddress.getByName().  It doesn't use anything fancy.

https://github.com/jboss-openshift/openshift-ping/blob/master/dns/src/main/java/org/openshift/ping/dns/GetServiceHosts.java#L19

It would be nice if that same logic could be used in the jgroups version, at least for A records.

**Comment by slaskawi** [comment]
> @luck3y, A records should work just fine to locate the pod endpoints, but you need to make sure the service is created with clusterIP: None. This is a "headless" service and the pod ip's should get served up for the A records (as opposed to the service ip, when the service is not headless).

Actually, there is no hard requirement for the Service to be headless. A cluster IP service should work as well. Please have a look at the Keycloak community template, where I proposed to use `DNS_PING` - https://github.com/jboss-dockerfiles/keycloak/pull/151/files#diff-c7387342b53ee5acf0c96f6b7f9191ebR230

> @slaskawi, not sure what you're asking. The openshift.DNS_PING implementation simply uses InetAddress.getByName(). It doesn't use anything fancy.

That's my point! The `DefaultDNSResolver` ([here's the code](https://github.com/belaban/JGroups/blob/master/src/org/jgroups/protocols/dns/DefaultDNSResolver.java)) queries DNS server through `DirContext`. From what you and @luck3y said, I have an impression that you would like to resolve DNS through the resolv mechanism in underlying OS. This approach would take the `/etc/resolv.conf` as well as `/etc/hosts' files into consideration. And as you said, we would need to use `InetAddress.getByName()` for it. So perhaps any of you would like to contribute another `DNSResolver` implementation, that will use `InetAddress.getByName()` to the `DNS_PING`? I just created a ticket for it - https://issues.jboss.org/browse/JGRP-2295

**Comment by rcernich** [comment]
@slaskawi thanks....if I find the time.

**Comment by rcernich** [comment]
@luck3y, are you saying this won't work in newer versions of java?

**Comment by luck3y** [comment]
@rcernich No I'm saying I think it should just work unmodified, I'll do some testing with nswitch / resolv.conf here and InetAddress.getByName(). If that's all thats needed, I can find the time to open a PR with the resolver change ;)

**Comment by slaskawi** [comment]
> Perhaps my understanding isn't correct, but I thought recent java runtimes used the underlying OS /etc/resolv.conf / nsswitch,conf now?

It depends. If you look carefully though the `DNS_PING` code, you will notice that there's no `InetAddress.getByName()` there :)

> @rcernich No I'm saying I think it should just work unmodified, I'll do some testing with nswitch / resolv.conf here and InetAddress.getByName(). If that's all thats needed, I can find the time to open a PR with the resolver change ;)

I'm a bit lost right now. Could you tell me what should work unmodified? do you refer to `dns.DNS_PING` or `openshift.DNS_PING`?

**Comment by slaskawi** [comment]
I already answered this question in
https://issues.jboss.org/browse/JGRP-2295. Thanks a lot Rob!

**Comment by luck3y** [comment]
> I'm a bit lost right now. Could you tell me what should work unmodified? do you refer to `dns.DNS_PING` or `openshift.DNS_PING`?

Sorry @slaskawi, I was unclear. I meant using InetAddress.getByName() should just work and honor the underlying OS settings in nsswitch / resolv.conf.

**Comment by luck3y** [comment]
FYI, I've updated this to seperate the KUBE_PING and DNS_PING config generation and untangle the hacks I was using for the namespace, which it now obtains from dns.DNS_PING namespace from either: OPENSHIFT_DNS_PING_SERVICE_NAMESPACE, or /var/run/secrets/kubernetes.io/serviceaccount/namespace.

I'd rather not have the env var around at all, but it makes it possible to test more easily, and possibly handles some unusual configurations.

As a reminder, openshift.DNS_PING will still remain the default in the templates, but the container will switch to using upstream kubernetes.KUBE_PING vs openshift.KUBE_PING as long as everything tests ok, so dns.DNS_PING we can have present and update to the changed upstream whenever that becomes available. (I think the query would probably be the same regardless.)

File: jboss-eap-config-jgroups/added/launch/ha.sh [hunk 4]
```diff
@@ -64,19 +72,18 @@ validate_dns_ping_settings() {
...
   declare protocol="$1"
-  if [ "${protocol}" = "openshift.KUBE_PING" ]; then
+  if [ "${protocol}" = "openshift.KUBE_PING" ] || [ "${protocol}" = "kubernetes.KUBE_PING" ]; then
     check_view_pods_permission
-  elif [ "${protocol}" = "openshift.DNS_PING" ]; then
+  elif [ "${protocol}" = "openshift.DNS_PING" ] || [ "${protocol}" = "dns.DNS_PING" ]; then
     validate_dns_ping_settings
   else
-    log_warning "Unknown protocol specified for JGroups discovery protocol: $1.  Expecting one of: openshift.KUBE_PING or openshift.DNS_PING."
+    log_warning "Unknown protocol specified for JGroups discovery protocol: $1. Expecting one of: kubernetes.KUBE_PING, openshift.KUBE_PING or openshift.DNS_PING."
```
[the remark this answers is not in the document]
Comment (luck3y) [reply jboss-eap-config-jgroups/added/launch/ha.sh]: Sorry @Crumby, I missed this one, thanks, will fix.

File: jboss-eap-config-jgroups/added/launch/ha.sh (hunk 1 above)
Comment (rcernich) [reply jboss-eap-config-jgroups/added/launch/ha.sh]: Ironically, I don't think we need the namespace at all for DNS.

Comment (luck3y) [reply jboss-eap-config-jgroups/added/launch/ha.sh]: @rcernich This was an early hacky version, it was removed later, and should be consistent now.

**Pull request merged**